src/client/datascience/notebook/vscodeNotebookController.ts

```diff
@@ -377,15 +369,31 @@ export class VSCodeNotebookController implements Disposable {
             .then(noop, (ex) => console.error(ex));
     }
 
-    private executeCell(doc: NotebookDocument, cell: NotebookCell) {
+    private async executeCell(doc: NotebookDocument, cell: NotebookCell) {
         traceInfo(`Execute Cell ${cell.index} ${getDisplayPath(cell.notebook.uri)}`);
-        const kernel = this.kernelProvider.getOrCreate(cell.notebook, {
-            metadata: this.kernelConnection,
-            controller: this.controller,
-            resourceUri: doc.uri
-        });
-        this.updateKernelInfoInNotebookWhenAvailable(kernel, doc);
-        return kernel.executeCell(cell);
+
+        // Start execution now (from the user's point of view)
+        const execution = CellExecutionCreator.getOrCreate(cell, this.controller);
+        execution.start(new Date().getTime());
+        void execution.clearOutput(cell);
+
+        // Connect to a matching kernel if possible (but user may pick a different one)
+        try {
+            const kernel = await this.connectToKernel(doc);
+            this.updateKernelInfoInNotebookWhenAvailable(kernel, doc);
+            return await kernel.executeCell(cell);
+        } catch (ex) {
+            // If there was a failure connecting or executing the kernel, stick it in this cell
+            return displayErrorsInCell(cell, execution, ex.toString());
+        }
+
+        // Execution should be ended elsewhere
+    }
+
+    @chainable()
+    private async connectToKernel(doc: NotebookDocument) {
+        // Make sure we don't have multiple cells trying to connect at the same time
```

Comment:
THis comment is confusing, we don't seem to doing anything with cells here, hence not sure about the comment

Reply:
executeCell gets called for all cells in a 'Run All' situation. This connectToKernel was chained so that only one of them would actually attempt to get the kernel at a time.

I'll change the comment to say that.